### PR TITLE
lint: add simple parenthesis check

### PR DESCRIPTION
**Pull request by fexolm**

### CODE REVIEW

**Comment by fexolm** [comment]
refs #9

File: lint/parenthesis-checker.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+)
+
+// ParenthesisChecker ...
```
Comment (Quasilyte) [inline lint/parenthesis-checker.go]: Don't be too lazy.
Either write at least proper sentence, or drop it completely and let linter rage continue (someone will fix that later).
You're only making it harder to refactor code later (fooling around linter).

File: testdata/foo.go [hunk 2]
```diff
@@ -11,3 +11,7 @@ func paramNamesCapitalized0(X, Y int) {}
 func paramNamesCapitalized1(X int) (Y, Z int) { return 0, 0 }
 
 func (PN paramNames) Capitalized2() {}
+
+func badReturn() [](func()) {
```
Comment (Quasilyte) [inline testdata/foo.go]: You can also add more complex tests like:
```go
func f() [](func([](func()))) {
	return nil
}
```
I think current implementation will only find outermost parenthesis and will not suggest optimal `[]func([]func())` form.

If it's this way, either add a TODO or make a note in referenced issue.
We can improve on this.

File: lint/parenthesis-checker.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+			continue
+		}
+		for _, res := range decl.Type.Results.List {
+			c.validateResultDecl(res)
+		}
+	}
+	return c.warnings
+}
+
+func (c *ParenthesisChecker) validateResultDecl(f *ast.Field) {
```
Comment (Quasilyte) [inline lint/parenthesis-checker.go]: We could work on this method design later to make it possible to re-use it for other AST nodes check (we don't want to be limited to results).

But for now, let's change `validate` to `check` for consistency (`checkResultDecl`).

File: lint/parenthesis-checker.go (hunk 1 above)
Comment (Quasilyte) [reply lint/parenthesis-checker.go]: It's actually very common bad practice to use `// Foo ...` or something like that.
We could add a checker for that.

**Comment by Quasilyte** [comment]
And by the way, good job. Made it ahead of other members. :p

File: lint/parenthesis-checker.go (hunk 1 above)
Comment (fexolm) [reply lint/parenthesis-checker.go]: thank you, i have created an issue for that #24

File: testdata/foo.go (hunk 2 above)
Comment (fexolm) [reply testdata/foo.go]: You are right, i had added TODO for that.

File: lint/parenthesis-checker.go [hunk 4]
```diff
@@ -5,19 +5,26 @@ import (
...
 
-// ParenthesisChecker ...
+// ParenthesisChecker detects some cases where parenthesis are unnecessary
 type ParenthesisChecker struct {
 	ctx *Context
 
 	warnings []Warning
 }
 
+// NewParenthesisChecker returns initialized checker for Go functions param names.
```
Comment (Quasilyte) [inline lint/parenthesis-checker.go]:
```diff
- NewParenthesisChecker returns initialized checker for Go functions param names.
+ NewParenthesisChecker returns initialized checker for type expressions.
```

Comment (Quasilyte) [reply lint/parenthesis-checker.go]: I think it will check more than function params eventually.

**Pull request merged**